internal/receiver/lightprometheusreceiver/scraper.go

```diff
@@ -0,0 +1,223 @@
+		}
+		return resp.Body, expfmt.ResponseFormat(resp.Header), nil
+	}
+	return s.fetchPrometheusMetrics(fetch)
+}
+
+func (s *scraper) fetchPrometheusMetrics(fetch fetcher) (pmetric.Metrics, error) {
+	metricFamilies, err := s.doFetch(fetch)
+	m := pmetric.NewMetrics()
+	if err != nil {
+		return m, err
+	}
+
+	u, err := url.Parse(s.cfg.Endpoint)
+	if err != nil {
+		return m, err
+	}
+	rm := m.ResourceMetrics().AppendEmpty()
+	res := rm.Resource()
+	res.Attributes().PutStr(conventions.AttributeServiceName, s.name)
+	res.Attributes().PutStr(conventions.AttributeNetHostName, u.Host)
+	res.Attributes().PutStr(conventions.AttributeServiceInstanceID, s.name)
+	res.Attributes().PutStr(conventions.AttributeNetHostPort, u.Port())
+	res.Attributes().PutStr(conventions.AttributeHTTPScheme, u.Scheme)
+	s.convertMetricFamilies(metricFamilies, rm)
+	return m, nil
+}
+
+func (s *scraper) doFetch(fetch fetcher) ([]*dto.MetricFamily, error) {
+	body, expformat, err := fetch()
+	if err != nil {
+		return nil, err
+	}
+	defer body.Close()
+	var decoder expfmt.Decoder
+	// some "text" responses are missing \n from the last line
+	if expformat != expfmt.FmtProtoDelim {
+		decoder = expfmt.NewDecoder(io.MultiReader(body, strings.NewReader("\n")), expformat)
+	} else {
+		decoder = expfmt.NewDecoder(body, expformat)
+	}
+
+	var mfs []*dto.MetricFamily
+
+	for {
+		var mf dto.MetricFamily
+		err := decoder.Decode(&mf)
+
+		if err == io.EOF {
+			return mfs, nil
+		} else if err != nil {
+			return nil, err
+		}
+
+		mfs = append(mfs, &mf)
+	}
+}
+
+func (s *scraper) convertMetricFamilies(metricFamilies []*dto.MetricFamily, rm pmetric.ResourceMetrics) {
+	now := pcommon.NewTimestampFromTime(time.Now())
+
+	sm := rm.ScopeMetrics().AppendEmpty()
+	for _, family := range metricFamilies {
+		newMetric := sm.Metrics().AppendEmpty()
+		newMetric.SetName(family.GetName())
+		newMetric.SetDescription(family.GetHelp())
+		switch *family.Type {
+		case dto.MetricType_COUNTER:
+			sum := newMetric.SetEmptySum()
+			sum.SetIsMonotonic(true)
+			sum.SetAggregationTemporality(pmetric.AggregationTemporalityCumulative)
+			for _, fm := range family.GetMetric() {
+				dp := sum.DataPoints().AppendEmpty()
+				dp.SetTimestamp(now)
+				dp.SetDoubleValue(fm.GetCounter().GetValue())
+				for _, l := range fm.GetLabel() {
+					if l.GetValue() != "" {
+						dp.Attributes().PutStr(l.GetName(), l.GetValue())
+					}
+				}
+			}
+		case dto.MetricType_GAUGE:
+			gauge := newMetric.SetEmptyGauge()
+			for _, fm := range family.Metric {
+				dp := gauge.DataPoints().AppendEmpty()
+				dp.SetDoubleValue(fm.GetGauge().GetValue())
+				dp.SetTimestamp(now)
+				for _, l := range fm.GetLabel() {
+					if l.GetValue() != "" {
+						dp.Attributes().PutStr(l.GetName(), l.GetValue())
+					}
+				}
+			}
+		case dto.MetricType_HISTOGRAM, dto.MetricType_GAUGE_HISTOGRAM:
```

Comment:
Would we consider adding an option to this receiver to drop certain type of prom metrics?
I have a use case where I want to collect prom metrics from some k8s components but am not really interested in the histogram metrics because the buckets are generated dynamically and w/o native UI support for histogram in O11y, visualizing these is not a good experience. So I would just like to not collect them instead of adding exclusion rules for specific metrics.

Reply:
good point, yes we can and I think we should. @atoulme wdyt?

Reply:
right, filtering at the source will save us a few bytes... anything we can do save one bit of memory at a time :-)

Reply:
let's do it in a separate PR. Less is more.

Reply:
https://signalfuse.atlassian.net/browse/OTL-2009 is filed to follow up.